Add capi byline tags

Reettaphant opened the pull request:
- Adds a byline picker and syncs bylines with composer
- The byline picker is specific to bylines at the moment but might be adapted to picking tags that don't allow text input if this doesn't make it too complicated
- All input fields are made bigger in this pr so that the tag picker fields won't look out of place (they need to be bigger to display tags)
- depends on: https://github.com/guardian/atom-maker/pull/28
<img width="1180" alt="screen shot 2017-06-08 at 12 50 08" src="https://user-images.githubusercontent.com/3066534/26927275-1e0a368c-4c49-11e7-970d-4aeba678e456.png">

mbarton commented on public/video-ui/src/components/FormFields/FormFieldBylinePicker.js:
```diff
@@ -0,0 +1,301 @@
+
+export default class FormFieldBylinePicker extends React.Component {
+  state = {
+    bylineTags: null,
```
Could just be an empty array?

mbarton commented on public/video-ui/src/components/FormFields/FormFieldBylinePicker.js:
```diff
@@ -0,0 +1,301 @@
+  };
+
+  updateInput = e => {
+    if (this.state.lastAction === 'SPACE') {
```
Rather than tracking keypresses, is it possible to write a pure function that goes from the input value to the state required?

Reettaphant replied:
We need a way of knowing if the user is deleting or adding content, I don't think there is a way of doing that without tracking the key presses.

mbarton commented on public/video-ui/src/util/tagParsers.js:
```diff
@@ -0,0 +1,31 @@
+          };
+        });
+      } else {
+        return new Promise(resolve => resolve(element));
```
Nitpick: can also be `Promise.resolve(element)`

mbarton commented on public/video-ui/src/util/removeStringTagDuplicates.js:
```diff
@@ -0,0 +1,37 @@
+export default function removeStringTagDuplicates(tag, tagValue) {
+  // Remove string input that appears in the last words of the tag input that is being added
```
Might be useful to provide a few example inputs and outputs in the comment?

jennysivapalan replied:
I think we should add a test!

Reettaphant replied:
I've added a few tests to this now.

mbarton commented on public/video-ui/src/components/FormFields/FormFieldBylinePicker.js:
```diff
@@ -0,0 +1,241 @@
+          this.setState({
+            bylineTags: bylines
+          });
+          return;
```
Is this required?

jennysivapalan commented on public/video-ui/src/components/FormFields/FormFieldBylinePicker.js:
```diff
@@ -0,0 +1,241 @@
+      const newValue = this.state.tagValue.slice(
+        0,
+        this.state.tagValue.length - 1
+      );
```
Can the parameters for `slice` be on one line? Might be more readable?

Reettaphant replied:
This is the pre-commit hook reorganising code. If it is difficult to read we should probably change how it works.

mbarton replied:
I'd be up for removing the line length rule, it often does this to me

Reettaphant commented:
@jennysivapalan @mbarton I've added tests for removing duplicate input and split some of the functionality in the byline picker to utilities functions. We can't get rid of listening to key presses because its the only way we can know when the user is deleting input and when they are adding input. The build should be passing once https://github.com/guardian/content-atom/pull/91 and https://github.com/guardian/atom-maker/pull/28 have been merged and new versions released.

akash1810 commented on public/video-ui/src/test/stringTagDuplicates.spec.js:
```diff
@@ -0,0 +1,56 @@
+const simpleTag = { id: 'tag', webTitle: 'tag' };
+
+describe('Removing string tag duplicates', () => {
+  it('does not trim when imput is empty', () => {
```
typo

akash1810 commented on public/video-ui/src/util/removeStringTagDuplicates.js:
```diff
@@ -0,0 +1,40 @@
+
+  if (firstTagMatch !== -1) {
+    numberOfMatches = 1;
+    for (var i = firstTagMatch + 1; i < tagWords.length; i++) {
```
could we use a `.forEach` here?

Reettaphant replied:
I want to break out of the loop immediately when a match is not found (e.g. if someone has typed in 'Mary Smith' but picks a tag with web title 'Mary Jones', we don't want to trim any words from the input string). I guess it could be done with a foreach as well, but I think this is simpler.

akash1810 commented on public/video-ui/styles/components/_forms.scss:
```diff
@@ -219,6 +247,91 @@
   }
 }
 
+.form__field__tags {
```
can be merged with the rules below

akash1810 commented on public/video-ui/styles/components/_forms.scss:
```diff
@@ -219,6 +247,91 @@
+
+
+    &:after {
+      content: '×';
```
minor: why not use the icons?

Reettaphant replied:
This is how atom workshop does multiselect elements, I think it is good to have consistent look on the pickers if possible.

Reettaphant commented on public/video-ui/src/components/FormFields/FormFieldBylinePicker.js:
```diff
@@ -0,0 +1,261 @@
+import { tagsFromStringList, tagsToStringList } from '../../util/tagParsers';
+import removeStringTagDuplicates from '../../util/removeStringTagDuplicates';
+
+export default class FormFieldBylinePicker extends React.Component {
```
I was going to change the whole name in the pr that makes this component a more generic tag picker so maybe it is enough to just do it then?

akash1810 commented on public/video-ui/src/components/VideoData/VideoData.js:
```diff
@@ -68,6 +69,16 @@ class VideoData extends React.Component {
+
+            <ManagedField
+              fieldLocation="byline"
+              name="Byline Tags"
```
could we make this `Byline` to be consistent with Composer?

akash1810 commented on public/video-ui/styles/components/_forms.scss:
```diff
@@ -219,6 +247,88 @@
+  width: 100%;
+  padding: 5px 5px;
+  cursor: pointer;
+  border-top: 12px solid $color650Grey;
```
we've 2 `border-top` rules now for the same class, can we delete one?
